doc/neps/ufunc-overrides.rst

```diff
@@ -153,49 +152,207 @@ tuple in the ``out`` keyword argument.
+the option for other classes to have :class:`ndarray` at the bottom of
+the type hierarchy.
+
+.. admonition:: Example
+
+   Type casting hierarchy
+
+   .. graphviz::
+
+      digraph array_ufuncs {
+         rankdir=BT;
+         A -> C;
+         B -> C;
+         D -> B;
+         ndarray -> A;
+         ndarray -> B;
+      }
+
+   The ``__array_ufunc__`` of type A can handle ndarrays, B can handle
+   ndarray and D, and C can handle A and B but not ndarrays or D.  The
+   result is a directed acyclic graph, and defines a type casting
+   hierarchy, with relations ``C > A > ndarray``, ``C > B > ndarray``,
+   ``C > B > D``. The type B is incompatible relative to A and vice
+   versa, and A and ndarray are incompatible relative to D.  Ufunc
+   expressions involving these classes should produce results of the
+   highest type involved or raise a :exc:`TypeError`.
 
 Subclass hierarchies
 --------------------
 
-Hierarchies of such containers (say, a masked quantity), are most easily
-constructed if methods consistently use :func:`super` to pass through
-the class hierarchy [7]_.  To support this, :class:`ndarray` has its own
-``__array_ufunc__`` method (which is equivalent to ``getattr(ufunc,
-method)(*inputs, **kwargs)``, i.e., if any of the (adjusted) inputs
-still defines ``__array_ufunc__`` that will be called in turn). This
-should be particularly useful for container-like subclasses of
-:class:`ndarray`, which add an attribute like a unit or mask to a
-regular :class:`ndarray`. Such classes can do possible adjustment of the
-arguments relevant to their own class, pass on to another class in the
-hierarchy using :func:`super` until the Ufunc is actually done, and then
-do possible adjustments of the outputs.
+Generally, it is desirable to mirror the class hierarchy in the ufunc
+type casting hierarchy. The recommendation is that an
+``__array_ufunc__`` implementation of a class should generally return
+`NotImplemented` unless the inputs are instances of the same class or
+superclasses.  This guarantees that in the type casting hierarchy,
+superclasses are below, subclasses above, and other classes are
+incompatible (sadly, the terminology for graphs and classes has reversed
+vertical sense).  Exceptions to this need to check they respect the
+implicit type casting hierarchy.
+
+Subclasses can be easily constructed if methods consistently use
+:func:`super` to pass through the class hierarchy [7]_.  To support
+this, :class:`ndarray` has its own ``__array_ufunc__`` method,
+equivalent to::
+
+    def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
+        # Cannot handle items that have __array_ufunc__ (other than our own).
+        outputs = kwargs.get('out', ())
+        for item in inputs + outputs):
+            if (hasattr(item, '__array_ufunc__') and
+                    type(item).__array_ufunc__ is not ndarray.__array_ufunc__):
+                return NotImplemented
```

Comment:
This doesn't do the superclass check discussed in the text above.
If that's dropped (and the default method deals with all subclasses), the text should argue how the type casting hierarchy stays valid.